core/store/src/trie/mem/flexible_data/encoding.rs

```diff
@@ -0,0 +1,132 @@
+use std::mem::MaybeUninit;
+
+use super::FlexibleDataHeader;
+
+/// Facilitates allocation and encoding of flexibly-sized data.
+pub struct RawEncoder {
+    data: Box<[u8]>,
+    pos: usize,
+}
+
+impl RawEncoder {
+    /// Creates a new memory allocation of the given size, returning an encoder
+    /// that can be used to initialize the allocated memory.
+    pub fn new(n: usize) -> RawEncoder {
+        let mut data = Vec::<u8>::new();
+        data.reserve_exact(n);
+        unsafe {
+            data.set_len(n);
+        }
+        RawEncoder { data: data.into_boxed_slice(), pos: 0 }
+    }
+
+    unsafe fn ptr(&mut self) -> *mut u8 {
+        self.data.as_mut_ptr().offset(self.pos as isize)
+    }
+
+    /// Encodes the given fixed-size field to the current encoder position,
+    /// and then advance the position by the size of the field.
+    pub unsafe fn encode<T>(&mut self, data: T) {
+        assert!(self.pos + std::mem::size_of::<T>() <= self.data.len());
+        (*(self.ptr() as *mut MaybeUninit<T>)).write(data);
+        self.pos += std::mem::size_of::<T>();
+    }
+
+    /// Encodes the given flexibly-sized part of the data to the current
+    /// encoder position, and then advance the position by the size of the
+    /// flexibly-sized part, as returned by `header.flexible_data_length()`.
+    pub unsafe fn encode_flexible<T: FlexibleDataHeader>(
+        &mut self,
+        header: &T,
+        data: T::InputData,
+    ) {
+        let length = header.flexible_data_length();
+        assert!(self.pos + length <= self.data.len());
+        header.encode_flexible_data(data, self.ptr());
+        self.pos += length;
+    }
+
+    /// Finishes the encoding process and returns a pointer to the allocated
+    /// memory. The caller is responsible for freeing the pointer later.
+    pub unsafe fn finish(self) -> *const u8 {
+        assert_eq!(self.pos, self.data.len());
+        Box::into_raw(self.data) as *const u8
+    }
+}
+
+/// Facilitates the decoding of flexibly-sized data.
+/// The lifetime 'a should be a lifetime that guarantees the availability of
+/// the memory being decoded from.
```

Comment:
If I understand the requirements for `'a` correctly, this should just be `data: &'a [u8]` (and probably `&'a [MaybeUninit<u8>]` as per the above)